[AAP-16160] Add RADIUS authenticator plugin

cutwater opened the pull request:
**WARNING: Do not squash**

cutwater committed:
Add RADIUS authenticator plugin

jctanner commented on ansible_base/authentication/backend.py:
```diff
@@ -35,6 +35,7 @@ def authenticate(self, request, *args, **kwargs):
         last_modified = None if last_modified_item is None else last_modified_item.get('modified')
 
         for authenticator_id, authenticator_object in get_authentication_backends(last_modified).items():
+            print(authenticator_id, authenticator_object)
```
leftover debug print?

cutwater replied:
Thanks, btw there is `flake8-print` plugin for such kind of issues and this check is also included into `ruff` linter tool.

john-westcott-iv commented on ansible_base/authentication/authenticator_plugins/radius.py:
```diff
@@ -0,0 +1,323 @@
+        return user
+
+
+# The code below is based on the source code of the django-radius library (https://github.com/robgolding/django-radius).
```
We already forked django-radius and use it in AWX here https://github.com/ansible/django-radius/blob/develop/radiusauth/backends/radius.py. Would it make sense to just modify that and use a git require? We could also do a PR to the actual django-radius as well.

cutwater replied:
Unless the Ansible fork is released to PyPI we have to stick to vendored code, because eventually DAB will have to be published there. And PyPI doesn't allow git dependencies.

john-westcott-iv replied:
FYI, longer term we also want to split out authenticators into their own repos and make a "pluggable" system like collections in Ansible. So that community members can contribute to individual plugins/write new ones.

john-westcott-iv reviewed (commented):
You had also mentioned testing with a side car. Can you add that logic into our side car configuration so that other devs can spin up an environment like you do for testing?

cutwater commented:
@john-westcott-iv I haven't used side car the official way, just deployed a container. Building an appropriate sidecar will take some additional time.

john-westcott-iv commented on ansible_base/authentication/authenticator_plugins/radius.py:
```diff
@@ -0,0 +1,323 @@
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+DICTIONARY = u"""
```
I really want to think about how we are including this file. I'm a little nervous because we took the source and then added changes. What do we do if that upstream changes? How does the next person know what was changed so they could re-apply the changes? Would it make sense to have a commit for the origional source and then a commit for the changes we made to said source? Or, if the change set is small enough, could we make a subclass of _RADIUSBackend and just override the functions we need to? Or maybe have a patch file which we could attempt to apply to a future version of the RADIUS file? I'm open to any other thoughts but I think we need to have something clearer to know "what we did".

cutwater replied:
I agree, I can make it multiple comments first importing original implementation and then adjusting it to our needs, so that we don't loose history.
Actually I would like to move the derived code to a separate file, so that it's easier to maintain those changes and separate the 3rd party code from our code. Can you please suggest where is the best place within our project structure to put this file?

john-westcott-iv replied:
I'm not sure it matters all to much at this time since we want to (eventually) split these out into their own repo. That being said, for the GitHub authenticators @jctanner needed a common configuration that he put in `authentication/authenticator_configurations`.  Since this is radius specific I might add something like `authentication/radius_backend` or something along those lines if you are going to have multiple files. If you don't want/need a full directory structure you could also put a single file into `utils`.

john-westcott-iv commented on ansible_base/authentication/authenticator_plugins/radius.py:
```diff
@@ -0,0 +1,323 @@
+        return user
+
+
+# The code below is based on the source code of the django-radius library (https://github.com/robgolding/django-radius).
```
Can we add a block comment explaining why we are pulling this in like this for future reference?

cutwater committed:
Vendoring the django-radius project

We have to include the django-radius [1] project source code in order to
implement RADIUS plugin for the Django Ansible Base project due to the
following reasons:

1. As of today the most recent release of the django-radius
   project is 1.5.0 (dated 2020-03-02), which depends on
   the `future` library, that is incompatible with Python 3.11.

   This dependency was removed in robgolding/django-radius#27,
   but it hasn't been released.

2. The `django-radius` library depends on global settings object, imported directly
   from `django.conf` module, while in DAB settings for authentication
   plugins are stored in the database.

3. Since this project will be published to PyPi, we cannot rely on the
   fork [2], maintained by the Ansible organization, because
   distributions pushed to PyPI must not have git dependencies.

The code in this commit is based on the commit 4c082470b034a318590c92e0f43467be92b0c885 [3]
in the django-radius repository and copied without modification.
All modifications will be introduced in the subsequent commits.

[1] https://github.com/robgolding/django-radius/
[2] https://github.com/ansible/django-radius/
[3] https://github.com/robgolding/django-radius/commit/4c082470b034a318590c92e0f43467be92b0c885

cutwater committed:
Downstream chnages to RADIUS backend

* Add support of local settings by using dependency injection.
  Settings object is passed to the RADIUSBackend class `__init__`
  method with a fallback to default django's global settings.
* Disable code formatting to minimize diff with upstream changes.
* Fix minor code issues.

john-westcott-iv commented on ansible_base/authentication/authenticator_plugins/radius/__init__.py:
```diff
@@ -0,0 +1,90 @@
+#  Copyright 2024 Red Hat, Inc.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
```
Lets take this out until we hear back on this.

john-westcott-iv reviewed (commented):
With this plugin in a directory instead of as a file does it show up in the results of [get_authenticator_plugins](https://github.com/ansible/django-ansible-base/blob/devel/ansible_base/authentication/authenticator_plugins/utils.py#L14)?

cutwater commented:
> With this plugin in a directory instead of as a file does it show up in the results of [get_authenticator_plugins](https://github.com/ansible/django-ansible-base/blob/devel/ansible_base/authentication/authenticator_plugins/utils.py#L14)?

Good point, so the `get_authenticator_plugins` doesn't support python packages, I see.

john-westcott-iv reviewed (approved):
Pending SQ unit tests and what not.

john-westcott-iv commented:
Can you run a manual test with a simple authenticator_map to ensure thats working properly?

cutwater commented:
Blocked by #280
